File: java/BinaryTree.java
// each node as at most two children 

public class BinaryTree {
  public class Node {
    int data;
    Node left, right;

    public Node(int item) {
      data = item;
      left = right = null;
    }
  }
  Node root;
  public BinaryTree() {
    root = null;
  }
  // function to insert new node with given key
  public void insert(int key) {
    root = insertRec(root, key);
  }
  // recursive function to insert a new node with key in BST
  Node insertRec(Node root, int key) {
    // if tree is empty, return new node
    if(root == null) {
      root = new Node(key);
      return root;
    }
    // otherwise, recur down tree
    if(key < root.data)
      root.left = insertRec(root.left, key);
    else if(key > root.data)
      root.right = insertRec(root.right, key);
    return root;
  }
  // utility function to print inorder traversal
  public void inorder() {
    inorderRec(root);
  }
  // utility function to do inorder traversal of the tree
  void inorderRec(Node root) {
    if(root != null) {
      inorderRec(root.left);
      System.out.println(root.data + " ");
      inorderRec(root.right);
    }
  }
}

File: c/binaryTree.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
  int data;
  struct Node *left, *right;
};

// function to create new node
struct Node* newNode(int data) {
  struct Node* node = (struct Node*)malloc(sizeof(struct Node));
  node->data = data;
  node->left = NULL;
  node->right = NULL;
  return node;
}
// function to insert a new node
void insert(struct Node** root, int data) {
  if(*root == NULL) {
    *root = newNode(data);
    return;
  }
  if(data < (*root)->data) {
    insert(&(*root)->right, data);
  }
}
// function to perform inorder traversal 
void inorderTraversal(struct Node* root) {
  if(root != NULL) {
    inorderTraversal(root->left);
    printf("%d ", root->data);
    inorderTraversal(root->right);
  }
}

// function to perform preorder traversal
void preorderTraversal(struct Node* root) {
  if(root != NULL) {
    printf("%d ", root->data);
    preorderTraversal(root->left);
    preorderTraversal(root->right);
  }
}
// function to perform postorder traversal
void postorderTraversal(struct Node* root) {
  if(root != NULL) {
    postorderTraversal(root->left);
    postorderTraversal(root->right);
    printf("%d ", root->data);
  }
}

int main() {
  struct Node* root = NULL;
  insert(&root, 50);
  insert(&root, 30);
  insert(&root, 20);
  insert(&root, 40);
  insert(&root, 70);
  insert(&root, 60);
  insert(&root, 80);


  printf("inorder traversal: ");
  inorderTraversal(root);
  printf("\n");

  printf("preorder traversal: ");
  preorderTraversal(root);
  printf("\n");

  printf("postorder traversal: ");
  postorderTraversal(root);
  printf("\n");

  return 0;
}
